subprojects/core/src/main/java/org/gradle/internal/buildtree/BuildInclusionCoordinator.java

```diff
@@ -62,70 +93,53 @@ public void prepareRootBuildForInclusion() {
 
     private void registerGlobalLibrarySubstitutions() {
         for (IncludedBuildState includedBuild : libraryBuilds) {
-            doRegisterSubstitutions(includedBuild);
+            withLockForBuild(includedBuild, () -> substitutionRegistry.registerSubstitutionsFor(includedBuild));
         }
     }
 
     public void registerSubstitutionsAvailableFor(BuildState build) {
         if (build instanceof RootBuildState) {
             registerGlobalLibrarySubstitutions();
         } else {
-            makeSubstitutionsAvailableFor(build, new HashSet<>());
+            withLockForBuild(build, () -> makeSubstitutionsAvailableFor(build));
         }
     }
 
     public void registerSubstitutionsProvidedBy(BuildState build) {
         if (build instanceof RootBuildState && registerRootSubstitutions) {
             // Make root build substitutions available
-            doRegisterSubstitutions((RootBuildState) build);
+            withLockForBuild(build, () -> substitutionRegistry.registerSubstitutionsFor((RootBuildState) build));
         }
     }
 
     public void prepareForPluginResolution(IncludedBuildState build) {
-        synchronized (registering) {
-            if (!registering.add(build)) {
-                return;
-            }
-            try {
-                build.ensureProjectsConfigured();
-                makeSubstitutionsAvailableFor(build, new HashSet<>());
-            } finally {
-                registering.remove(build);
-            }
-        }
+        withLockForBuild(build, build::ensureProjectsConfigured);
+        makeSubstitutionsAvailableFor(build);
```

Comment:
`makeSubstitutionsAvailableFor` is not under the `build` lock to untie the things. Don't have 100% confidence tho